### DISCUSSION TITLE
API/Functions for Memory Management for Observable Counters upon Process Termination

### DISCUSSION THREAD

**Discussion by IamVishalSharma** [open-telemetry/opentelemetry-cpp#2524] [General]
- We're using Observable counters, numbering in the hundreds. They constantly gather and send metrics. 
- We typically run valgrind on our processes. When the process ends, we're noticing a bunch of memory allocation stacks related to these counters. This is normal, 
- For a clean shutdown, we need an API to stop this data collection and free up the memory. 
- We tried setting the provider to none as mentioned in example code of git, but it didn't work. 
- We couldn't find any helpful functions in the API either. The best we found was a way to deregister callback functions.

Please share your inputs

**Comment by lalitb** [comment]
@IamVishalSharma  Can you try calling `MeterProvider::RemoveMeter()` API before `MeterProvider::Shutdown()`. Also, release the ownership of the Meter's shared_ptr got with `MeterProvider::GetMeter()`. By releasing the shared ownership of the Meter, and assuming no other shared references are held elsewhere, the Meter and all associated objects (Instruments, Aggregates, etc.) will be destructed. Sequence would be something like:
```cpp
  auto meter = provider->GetMeter(name, version, url);
  ...
  ..
  //cleanup
  provider->RemoveMeter(name, version, url);
  meter->reset();
  provider->Shutdown();
```

**Reply by lalitb** [reply]
Forgot to mention, `MeterProvider::RemoveMeter()` is part of ABIv2, so you need to set OPENTELEMETRY_ABI_VERSION_NO macro to 2 during build.

**Reply by IamVishalSharma** [reply]
Thank you very much for your inputs @lalitb I will try these and let you know the outcome.

**Reply by varkey98** [reply]
But `provider->Shutdown()` is not part of the api right?

**Reply by lalitb** [reply]
Yes you are correct. Can you look into these changes if it helps - https://github.com/open-telemetry/opentelemetry-cpp/pull/2376
So 
 - If the ObservableCounter is destroyed, it's collection is stopped.
 - The memory allocated for aggregated Metrics is not cleaned-up. The specs doesn't talk much about it, but probably we need to provide some solution for this in future.

**Reply by varkey98** [reply]
@lalitb What should we do for normal counters? I have started a discussion here: https://github.com/open-telemetry/opentelemetry-cpp/discussions/2665

**Reply by varkey98** [reply]
Does that mean we dont have a cleanup method at all currently for the metrics? Also since we have the Shutdown function defined publicly in the sdk headers, would it be made part of the api in the future?

**Reply by lalitb** [reply]
> Does that mean we dont have a cleanup method at all currently for the metrics?

Cleanup should happen automatically once the instrument is destroyed. None of the otel language implementations do it as of now. So yes to your question as of now.

> Also since we have the Shutdown function defined publicly in the sdk headers, would it be made part of the api in the future?

No, API (which is used by libraries) should never call shutdown.

> What should we do for normal counters?

You can just let the synchronous instrument get destroyed(or go out of scope). Cleanup is not implemented for them too.

**Reply by varkey98** [reply]
Thanks for the clarification
